File: learning-factory/warehouse/artifacts/projects/build-your-own-x/remediation/54fbf20b587133619a1f/repair-v1/job_byox_repair_v1_g1_250e5b6563a90a708220b0592e8a0dfe/attempt-001/sealed/reference/lexer.js
import { LexerError } from "./errors.js";

export const TokenType = Object.freeze({
  LEFT_PAREN: "LEFT_PAREN",
  RIGHT_PAREN: "RIGHT_PAREN",
  LEFT_BRACE: "LEFT_BRACE",
  RIGHT_BRACE: "RIGHT_BRACE",
  SEMICOLON: "SEMICOLON",
  PLUS: "PLUS",
  MINUS: "MINUS",
  STAR: "STAR",
  SLASH: "SLASH",
  BANG: "BANG",
  BANG_EQUAL: "BANG_EQUAL",
  EQUAL: "EQUAL",
  EQUAL_EQUAL: "EQUAL_EQUAL",
  GREATER: "GREATER",
  GREATER_EQUAL: "GREATER_EQUAL",
  LESS: "LESS",
  LESS_EQUAL: "LESS_EQUAL",
  IDENTIFIER: "IDENTIFIER",
  NUMBER: "NUMBER",
  LET: "LET",
  SET: "SET",
  EMIT: "EMIT",
  IF: "IF",
  ELSE: "ELSE",
  WHILE: "WHILE",
  TRUE: "TRUE",
  FALSE: "FALSE",
  EOF: "EOF",
});

const KEYWORDS = Object.freeze({
  let: TokenType.LET,
  set: TokenType.SET,
  emit: TokenType.EMIT,
  if: TokenType.IF,
  else: TokenType.ELSE,
  while: TokenType.WHILE,
  true: TokenType.TRUE,
  false: TokenType.FALSE,
});

const SINGLE_CHARACTER_TYPES = Object.freeze({
  "(": TokenType.LEFT_PAREN,
  ")": TokenType.RIGHT_PAREN,
  "{": TokenType.LEFT_BRACE,
  "}": TokenType.RIGHT_BRACE,
  ";": TokenType.SEMICOLON,
  "+": TokenType.PLUS,
  "-": TokenType.MINUS,
  "*": TokenType.STAR,
  "/": TokenType.SLASH,
});

function isDigit(character) {
  return character >= "0" && character <= "9";
}

function isIdentifierStart(character) {
  return (character >= "a" && character <= "z")
    || (character >= "A" && character <= "Z")
    || character === "_";
}

function isIdentifierPart(character) {
  return isIdentifierStart(character) || isDigit(character);
}

/** Convert Pebble source text into position-bearing public token records. */
export function tokenize(source) {
  if (typeof source !== "string") {
    throw new TypeError("tokenize expects source text");
  }

  const tokens = [];
  let offset = 0;
  let line = 1;
  let column = 1;

  const peek = (distance = 0) => source[offset + distance] ?? "\0";
  const advance = () => {
    const character = source[offset] ?? "\0";
    offset += 1;
    if (character === "\n") {
      line += 1;
      column = 1;
    } else {
      column += 1;
    }
    return character;
  };
  const add = (type, lexeme, literal, start) => {
    tokens.push({ type, lexeme, literal, line: start.line, column: start.column });
  };

  while (offset < source.length) {
    const character = peek();

    if (character === " " || character === "\t" || character === "\r" || character === "\n") {
      advance();
      continue;
    }

    if (character === "/" && peek(1) === "/") {
      while (offset < source.length && peek() !== "\n") advance();
      continue;
    }

    const start = { line, column, offset };

    if (isDigit(character)) {
      while (isDigit(peek())) advance();
      if (peek() === "." && isDigit(peek(1))) {
        advance();
        while (isDigit(peek())) advance();
      }
      const lexeme = source.slice(start.offset, offset);
      const literal = Number(lexeme);
      if (!Number.isFinite(literal)) {
        throw new LexerError("numeric literal is not finite", start, "INVALID_NUMBER");
      }
      add(TokenType.NUMBER, lexeme, literal, start);
      continue;
    }

    if (isIdentifierStart(character)) {
      while (isIdentifierPart(peek())) advance();
      const lexeme = source.slice(start.offset, offset);
      const type = Object.hasOwn(KEYWORDS, lexeme) ? KEYWORDS[lexeme] : TokenType.IDENTIFIER;
      const literal = type === TokenType.TRUE ? true : type === TokenType.FALSE ? false : null;
      add(type, lexeme, literal, start);
      continue;
    }

    const singleType = SINGLE_CHARACTER_TYPES[character];
    if (singleType !== undefined) {
      add(singleType, advance(), null, start);
      continue;
    }

    const twoCharacter = `${character}${peek(1)}`;
    const pairedType = {
      "!=": TokenType.BANG_EQUAL,
      "==": TokenType.EQUAL_EQUAL,
      ">=": TokenType.GREATER_EQUAL,
      "<=": TokenType.LESS_EQUAL,
    }[twoCharacter];
    if (pairedType !== undefined) {
      advance();
      advance();
      add(pairedType, twoCharacter, null, start);
      continue;
    }

    const operatorType = {
      "!": TokenType.BANG,
      "=": TokenType.EQUAL,
      ">": TokenType.GREATER,
      "<": TokenType.LESS,
    }[character];
    if (operatorType !== undefined) {
      add(operatorType, advance(), null, start);
      continue;
    }

    throw new LexerError(`unexpected character ${JSON.stringify(character)}`, start);
  }

  tokens.push({ type: TokenType.EOF, lexeme: "", literal: null, line, column });
  return tokens;
}

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/from-source-code-to-machine-code-build-your-own-compiler-from-scratch-2877aeded6/job_byox_build_v1_7a07ec294de1cc10739ceacd586b1ffd/attempt-001/sealed/reference/minnow/lexer.py
"""Deterministic ASCII-token lexer with Unicode input rejection."""

from .errors import LexError
from .model import Token, TokenKind


MAX_I64 = (1 << 63) - 1

KEYWORDS = {
    "let": TokenKind.LET,
    "print": TokenKind.PRINT,
    "if": TokenKind.IF,
    "else": TokenKind.ELSE,
    "while": TokenKind.WHILE,
    "true": TokenKind.TRUE,
    "false": TokenKind.FALSE,
}

SINGLE = {
    "+": TokenKind.PLUS,
    "-": TokenKind.MINUS,
    "*": TokenKind.STAR,
    "/": TokenKind.SLASH,
    "%": TokenKind.PERCENT,
    ";": TokenKind.SEMICOLON,
    "(": TokenKind.LEFT_PAREN,
    ")": TokenKind.RIGHT_PAREN,
    "{": TokenKind.LEFT_BRACE,
    "}": TokenKind.RIGHT_BRACE,
}

PREFIXED = {
    "!": (TokenKind.BANG, TokenKind.BANG_EQUAL),
    "=": (TokenKind.EQUAL, TokenKind.EQUAL_EQUAL),
    "<": (TokenKind.LESS, TokenKind.LESS_EQUAL),
    ">": (TokenKind.GREATER, TokenKind.GREATER_EQUAL),
}


def _is_alpha(character):
    return "A" <= character <= "Z" or "a" <= character <= "z" or character == "_"


def _is_digit(character):
    return "0" <= character <= "9"


class Lexer:
    def __init__(self, source):
        self.source = source
        self.index = 0
        self.line = 1
        self.column = 1
        self.tokens = []

    def scan(self):
        while True:
            self._skip_trivia()
            if self._at_end():
                self.tokens.append(Token(TokenKind.EOF, "", self.line, self.column))
                return self.tokens
            self._scan_token()

    def _at_end(self):
        return self.index >= len(self.source)

    def _peek(self, distance=0):
        position = self.index + distance
        if position >= len(self.source):
            return "\0"
        return self.source[position]

    def _take_plain(self):
        character = self.source[self.index]
        self.index += 1
        self.column += 1
        return character

    def _take_newline(self):
        if self._peek() == "\r":
            self.index += 1
            if self._peek() == "\n":
                self.index += 1
        else:
            self.index += 1
        self.line += 1
        self.column = 1

    def _skip_trivia(self):
        while not self._at_end():
            character = self._peek()
            if character in " \t":
                self._take_plain()
            elif character in "\r\n":
                self._take_newline()
            elif character == "/" and self._peek(1) == "/":
                self._take_plain()
                self._take_plain()
                while not self._at_end() and self._peek() not in "\r\n":
                    self._take_plain()
            else:
                return

    def _scan_token(self):
        start = self.index
        line = self.line
        column = self.column
        character = self._take_plain()

        if _is_digit(character):
            while _is_digit(self._peek()):
                self._take_plain()
            lexeme = self.source[start:self.index]
            value = int(lexeme, 10)
            if value > MAX_I64:
                raise LexError("integer literal exceeds signed 64-bit maximum", line=line, column=column, code="LEX002")
            self.tokens.append(Token(TokenKind.INT, lexeme, line, column, value))
            return

        if _is_alpha(character):
            while _is_alpha(self._peek()) or _is_digit(self._peek()):
                self._take_plain()
            lexeme = self.source[start:self.index]
            self.tokens.append(Token(KEYWORDS.get(lexeme, TokenKind.IDENT), lexeme, line, column))
            return

        if character in PREFIXED:
            one, two = PREFIXED[character]
            kind = one
            if self._peek() == "=":
                self._take_plain()
                kind = two
            self.tokens.append(Token(kind, self.source[start:self.index], line, column))
            return

        if character in SINGLE:
            self.tokens.append(Token(SINGLE[character], character, line, column))
            return

        raise LexError(f"unexpected character {character!r}", line=line, column=column)


def tokenize(source):
    if not isinstance(source, str):
        raise TypeError("source must be str")
    return Lexer(source).scan()
